### PR TITLE
[Planning Pipeline Refactoring] #2 Enable chaining planners

### PR DESCRIPTION
**Pull request by sjahr**
Related PRs:
- https://github.com/ros-planning/moveit_resources/pull/191
- https://github.com/ros-planning/moveit2_tutorials/pull/820

### Description

This PR enables chaining multiple planners. This way it is e.g. possible to create an initial trajectory or reference trajectory for optimizing planners.

### TODO
- [ ] Add unittest for getTrajectoryConstraints
- [ ] Write issue to adapt MotionPlanRequest to changes
- [ ] Write issue to make chain respect max. planning time
- [ ] Write issue to update move_group interface
- [ ] Test if Benchmarking still work

### Testing
-  Use https://github.com/ros-planning/moveit2_tutorials/pull/820 and https://github.com/ros-planning/moveit_resources/pull/191
- Update your *_planning.yaml files like this:
```diff
- planner_plugin: my_fancy/Planner1
+ planner_plugins:
+  my_fancy/Planner1 # for example ompl_interface/OMPLPlanner
+  my_fancy/Planner2 # for example stomp_moveit/StompPlanner
```
- run `ros2 launch moveit2_tutorials pipeline_testbench.launch.py`
### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit2/blob/main/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://moveit.picknik.ai/humble/doc/examples/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_configs_utils/default_configs/chomp_planning.yaml [hunk 1]
```diff
@@ -1,4 +1,5 @@
-planning_plugin: chomp_interface/CHOMPPlanner
+planning_plugin:
```
Comment (sea-bass) [inline moveit_configs_utils/default_configs/chomp_planning.yaml:1]: typo

```suggestion
planning_plugins:
```

File: moveit_ros/benchmarks/include/moveit/benchmarks/BenchmarkExecutor.h [hunk 2]
```diff
@@ -170,8 +170,7 @@ class BenchmarkExecutor
   void shiftConstraintsByOffset(moveit_msgs::msg::Constraints& constraints, const std::vector<double>& offset);
 
   /// Check that the desired planner plugins and algorithms exist for the given group
-  bool plannerConfigurationsExist(const std::map<std::string, std::vector<std::string>>& planners,
-                                  const std::string& group_name);
+  bool plannerConfigurationsExist(const std::map<std::string, std::vector<std::string>>& planners);
```
Comment (sea-bass) [inline moveit_ros/benchmarks/include/moveit/benchmarks/BenchmarkExecutor.h:173]: Why did we take off the group name? And either way, the docstring then needs to be updated?

File: moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp [hunk 3]
```diff
@@ -73,94 +73,98 @@ void MoveGroupQueryPlannersService::initialize()
...
-      planner_interface->getPlanningAlgorithms(algs);
-      moveit_msgs::msg::PlannerInterfaceDescription pi_desc;
-      pi_desc.name = planner_interface->getDescription();
-      pi_desc.pipeline_id = pipeline_id;
-      planner_interface->getPlanningAlgorithms(pi_desc.planner_ids);
-      res->planner_interfaces.push_back(pi_desc);
-    }
-  }
-  return true;
+  // TODO(sjahr): This is currently not working. Need to decide whether to fix or remove this.
```
Comment (sea-bass) [inline moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp:78]: Where is this used currently?

File: moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp [hunk 4]
```diff
@@ -73,94 +73,98 @@ void MoveGroupQueryPlannersService::initialize()
...
-    }
-
-    for (const auto& key_value_pair : config)
-    {
-      res->params.keys.push_back(key_value_pair.first);
-      res->params.values.push_back(key_value_pair.second);
-    }
-  }
-  return true;
+  // TODO(sjahr): This is currently not working. Need to decide whether to fix or remove this.
```
Comment (sea-bass) [inline moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp:103]: same here, where is this and `setParams()` used currently?

File: moveit_ros/planning/planning_pipeline/include/moveit/planning_pipeline/planning_pipeline.h [hunk 5]
```diff
@@ -163,6 +162,14 @@ class MOVEIT_PLANNING_PIPELINE_EXPORT PlanningPipeline
   {
     return false;
   }
+  [[deprecated("Please use getPlannerPluginNames().")]] const std::string& getPlannerPluginName() const
+  {
+    return pipeline_parameters_.planning_plugins.at(0);
+  }
+  [[deprecated("Removed from API.")]] const planning_interface::PlannerManagerPtr& getPlannerManager()
```
Comment (sea-bass) [inline moveit_ros/planning/planning_pipeline/include/moveit/planning_pipeline/planning_pipeline.h:169]: Is there no further rationale for this being removed from the API that we can display?

File: moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp (hunk 3 above)
Comment (sjahr) [reply moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp:78]: Never mind, looks like the RVIZ plugins need this :see_no_evil:

File: moveit_ros/planning/planning_pipeline/include/moveit/planning_pipeline/planning_pipeline.h (hunk 5 above)
Comment (sjahr) [reply moveit_ros/planning/planning_pipeline/include/moveit/planning_pipeline/planning_pipeline.h:169]: I am hesitant if raw access to the planner itself should be provided by the pipeline class :thinking: Looks like the main reason for that has been so you can read configs from the planner but that could be done more explicitly

File: moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp (hunk 4 above)
Comment (sjahr) [reply moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp:103]: See above. This needs to be updated

File: moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp [hunk 6]
```diff
@@ -49,118 +50,153 @@ void MoveGroupQueryPlannersService::initialize()
...
   if (!planning_pipeline)
-    return false;
+  {
+    RCLCPP_ERROR(moveit::getLogger(), "Pipeline '%s' does not exist.", req->pipeline_id.c_str());
+    return;
+  }
 
-  const planning_interface::PlannerManagerPtr& planner_interface = planning_pipeline->getPlannerManager();
-  if (planner_interface)
+  // TODO(sjahr): Update for multiple planner plugins
```
Comment (sea-bass) [inline moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp:111]: Does this need to get done, or is it updated?

File: moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp [hunk 7]
```diff
@@ -49,118 +50,153 @@ void MoveGroupQueryPlannersService::initialize()
...
   const planning_pipeline::PlanningPipelinePtr planning_pipeline = resolvePlanningPipeline(req->pipeline_id);
   if (!planning_pipeline)
-    return false;
+  {
+    RCLCPP_ERROR(moveit::getLogger(), "Pipeline '%s' does not exist.", req->pipeline_id.c_str());
+    return;
+  }
 
-  const planning_interface::PlannerManagerPtr& planner_interface = planning_pipeline->getPlannerManager();
+  // TODO(sjahr): Update for multiple planner plugins
```
Comment (sea-bass) [inline moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp:168]: same here, is this TODO done now?

File: moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp (hunk 6 above)
Comment (sjahr) [reply moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp:111]: This needs to be updated. For now the first planner in the chain is used which is usually OMPL. Updating this involves more changes in the visualization with RVIZ and the moveit plugins which is why I'd like to keep this out of the PR

File: moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp (hunk 7 above)
Comment (sjahr) [reply moveit_ros/move_group/src/default_capabilities/query_planners_service_capability.cpp:168]: See my comment above

File: moveit_ros/planning/planning_pipeline/src/planning_pipeline.cpp [hunk 8]
```diff
@@ -241,22 +284,38 @@ bool PlanningPipeline::generatePlan(const planning_scene::PlanningSceneConstPtr&
...
-      planning_interface::PlanningContextPtr context =
-          planner_instance_->getPlanningContext(planning_scene, mutable_request, res.error_code);
-      if (context)
+      const auto& planner = planner_map_.at(planner_name);
+      // Update reference trajectory with latest solution (if available)
+      if (res.trajectory)
       {
-        context->solve(res);
-        publishPipelineState(mutable_request, res, planner_instance_->getDescription());
+        mutable_request.trajectory_constraints.constraints = getTrajectoryConstraints(res.trajectory);
```
Comment (sjahr) [inline moveit_ros/planning/planning_pipeline/src/planning_pipeline.cpp:294]: @sea-bass

**Review by tylerjw — APPROVED** [review]

**Comment by Betavion** [comment]
When chaining the planners (ompl & chomp), chomp can't seem to obtain ompl's planning result.
So chomp has no seed path to optimize.

My config file `chomp_planning.yaml` looks like this:
``` yaml
request_adapters:
  - default_planning_request_adapters/ResolveConstraintFrames
  - default_planning_request_adapters/ValidateWorkspaceBounds
  - default_planning_request_adapters/CheckStartStateBounds
  - default_planning_request_adapters/CheckStartStateCollision
planning_plugins:
  - ompl_interface/OMPLPlanner
  - chomp_interface/CHOMPPlanner
response_adapters:
  - default_planning_response_adapters/AddTimeOptimalParameterization
  - default_planning_response_adapters/ValidateSolution
  - default_planning_response_adapters/AddRuckigTrajectorySmoothing
  - default_planning_response_adapters/DisplayMotionPath
planning_time_limit: 10.0
max_iterations: 200
max_iterations_after_collision_free: 5
smoothness_cost_weight: 0.1
obstacle_cost_weight: 1.0
learning_rate: 0.01
animate_path: true
add_randomness: false
smoothness_cost_velocity: 0.0
smoothness_cost_acceleration: 1.0
smoothness_cost_jerk: 0.0
hmc_discretization: 0.01
hmc_stochasticity: 0.01
hmc_annealing_factor: 0.99
use_hamiltonian_monte_carlo: false
ridge_factor: 0.0
use_pseudo_inverse: false
pseudo_inverse_ridge_factor: 0.0001
animate_endeffector: false
animate_endeffector_segment: "tool"
joint_update_limit: 0.1
collision_clearance: 0.2
collision_threshold: 0.07
random_jump_amount: 1.0
use_stochastic_descent: true
enable_failure_recovery: false
max_recovery_attempts: 5
trajectory_initialization_method: "fillTrajectory"
```

The error:
``` bash
[rviz2-3] [INFO] [1752514314.532158177] [interactive_marker_display_98252816441568]: Sending request for interactive markers
[rviz2-3] [INFO] [1752514314.564684918] [interactive_marker_display_98252816441568]: Service response received for initialization
[move_group-2] [INFO] [1752514327.607457599] [move_group.moveit.moveit.ros.move_group.move_action]: MoveGroupMoveAction: Received request
[rviz2-3] [INFO] [1752514327.607594257] [moveit_2512680583.moveit.ros.move_group_interface]: Plan and Execute request accepted
[move_group-2] [INFO] [1752514327.607681415] [move_group.moveit.moveit.ros.move_group.move_action]: executing..
[move_group-2] [INFO] [1752514327.615760678] [move_group.moveit.moveit.ros.move_group.move_action]: Combined planning and execution request received for MoveGroup action. Forwarding to planning and execution pipeline.
[move_group-2] [INFO] [1752514327.615946455] [move_group.moveit.moveit.ros.plan_execution]: Planning attempt 1 of at most 1
[move_group-2] [INFO] [1752514327.616010161] [move_group.moveit.moveit.ros.move_group.capability]: Using planning pipeline 'chomp'
[move_group-2] [INFO] [1752514327.616277576] [move_group]: Calling PlanningRequestAdapter 'ResolveConstraintFrames'
[move_group-2] [INFO] [1752514327.616307480] [move_group]: Calling PlanningRequestAdapter 'ValidateWorkspaceBounds'
[move_group-2] [INFO] [1752514327.616314393] [move_group]: Calling PlanningRequestAdapter 'CheckStartStateBounds'
[move_group-2] [INFO] [1752514327.616330051] [move_group]: Calling PlanningRequestAdapter 'CheckStartStateCollision'
[move_group-2] [INFO] [1752514327.616920722] [move_group.moveit.moveit.planners.ompl.model_based_planning_context]: Planner configuration 'sia_arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
[move_group-2] [INFO] [1752514327.617076024] [move_group]: Calling Planner 'OMPL'
[move_group-2] [INFO] [1752514327.869064763] [move_group]: Calling Planner 'CHOMP'
[move_group-2] [ERROR] [1752514327.869187997] [move_group.moveit.moveit.planners.chomp.planner]: No input trajectory specified
[move_group-2] [ERROR] [1752514327.869244029] [move_group]: Planner 'CHOMP' failed with error code UNDEFINED
[move_group-2] [INFO] [1752514327.869312843] [move_group.moveit.moveit.ros.move_group.move_action]: UNDEFINED
[rviz2-3] [INFO] [1752514327.869681041] [moveit_2512680583.moveit.ros.move_group_interface]: Plan and Execute request aborted
[rviz2-3] [ERROR] [1752514327.870363289] [moveit_2512680583.moveit.ros.move_group_interface]: MoveGroupInterface::move() failed or timeout reached
```
